crates/shrs_core/src/shell.rs

```diff
@@ -49,14 +49,18 @@ impl Shell {
 
     // Trigger a hook of given type with payload
     pub fn run_hooks<C: HookCtx>(&self, c: C) {
-        self.cmd.run(move |sh: &mut Shell, states: &States| {
+        self.cmd.run(move |sh: &mut Shell, states: &mut States| {
             let _ = sh.hooks.run(sh, states, &c);
             sh.apply_queue(states);
         })
     }
+    pub(crate) fn run_hooks_in_core<C: HookCtx>(&mut self, states: &mut States, c: C) {
+        self.run_hooks(c);
```

Comment:
doesnt this cause hooks to be applied twice? since there is an `apply_queue` call inside the closure of `run_hooks`

Reply:
so this line should be `let _ = sh.hooks.run(sh, states, &c);` instead? Let's also rename this function to `_run_hooks` or something and call it in `run_hooks`

Reply:
apply_queue consumes everything so thats not possible. The issue is self.cmd.run only queues up the hook_run if u dont call apply_queue afterwards to start the chain.

Reply:
run_hooks also doesnt take in states so it cant call apply queue